hw6/oop_2.py

```diff
@@ -0,0 +1,157 @@
+    (нужно гаранитировать отсутствие повторяющихся результатов по каждому
+    заданию), группировать по экземплярам Homework.
+    Общий для всех учителей. Вариант ипользования смотри в блоке if __main__...
+Методы:
+    check_homework - принимает экземпляр HomeworkResult и возвращает True если
+    ответ студента больше 5 символов, так же при успешной проверке добавить в
+    homework_done.
+    Если меньше 5 символов - никуда не добавлять и вернуть False.
+    reset_results - если передать экземпряр Homework - удаляет только
+    результаты этого задания из homework_done, если ничего не передавать,
+    то полностью обнулит homework_done.
+PEP8 соблюдать строго.
+Всем перечисленным выше атрибутам и методам классов сохранить названия.
+К названием остальных переменных, классов и тд. подходить ответственно -
+давать логичные подходящие имена.
+"""
+from collections import defaultdict
+from datetime import datetime, timedelta
+
+
+class DeadlineError(Exception):
+    """ Error for situation when have done hw too late"""
+
+    pass
+
+
+class Person:
+    def __init__(self, first_name, last_name):
+        self.first_name = first_name
+        self.last_name = last_name
+
+
+class Homework:
+    def __init__(self, text: str, deadline: timedelta):
+        self.text = text
+        self.deadline = deadline
+        self.created = datetime.now()
+
+    def is_active(self):
+        """
+        Method for checking whether homework is expired or not
+        """
+        current_data = datetime.now()
+        return current_data - self.created < self.deadline
+
+
+class Student(Person):
+    def __init__(self, first_name: str, last_name: str):
+        super(Student, self).__init__(first_name, last_name)
+
+    def do_homework(self, homework: Homework, solution: str):
+        """
+        Return a HomeworkResult instance if homework isn't expired
+        """
+        if not homework.is_active():
+            raise DeadlineError("You are late")
+        return HomeworkResult(homework, solution, author=self, created=datetime.now())
+
+
+class HomeworkResult:
+    def __init__(
+        self, homework: Homework, solution: str, author: Student, created: datetime
+    ):
+        if not isinstance(homework, Homework):
+            raise ValueError("You gave not a Homework object")
+        self.homework = homework
+        self.solution = solution
+        self.author = author
+        self.created = created
+
+
+class Teacher(Person):
+    homework_done = defaultdict(set)  # what to do there?
+
+    def __init__(self, first_name: str, last_name: str):
+        super().__init__(first_name, last_name)
+
+    def check_homework(self, hw_result: HomeworkResult):
+        """
+        Add HomeworkResult instance to homework_done if solution is valid
+        """
+        if len(hw_result.solution) < 5:
+            return False
+        type(self).homework_done[hw_result.homework].add(hw_result)
```

Comment:
how can we get rid of `type` here?

Reply:
We can write either self.__class__ or just Teacher. The first one is more preferable for me because I can copy this method and use it in another class for example. So I changed it to self.__class__ - Done.